cdc/sinkv2/eventsink/txn/mysql/mysql.go

```diff
@@ -214,33 +378,49 @@ func (s *mysqlBackend) prepareDMLs() *preparedDMLs {
 	// translateToInsert control the update and insert behavior
 	// we only translate into insert when old value is enabled and safe mode is disabled
 	translateToInsert := s.cfg.EnableOldValue && !s.cfg.SafeMode
-	for _, event := range s.events {
-		for _, row := range event.Event.Rows {
-			if !translateToInsert {
-				break
-			}
-			// A row can be translated in to INSERT, when it was committed after
-			// the table it belongs to been replicating by TiCDC, which means it must not be
-			// replicated before, and there is no such row in downstream MySQL.
-			translateToInsert = row.CommitTs > row.ReplicatingTs
-		}
-	}
 
 	rowCount := 0
 	for _, event := range s.events {
+		if len(event.Event.Rows) == 0 {
+			continue
+		}
+
+		firstRow := event.Event.Rows[0]
+		if len(startTs) == 0 || startTs[len(startTs)-1] != firstRow.StartTs {
+			startTs = append(startTs, firstRow.StartTs)
+		}
+
+		// A row can be translated in to INSERT, when it was committed after
+		// the table it belongs to been replicating by TiCDC, which means it must not be
+		// replicated before, and there is no such row in downstream MySQL.
+		translateToInsert = translateToInsert && firstRow.CommitTs > firstRow.ReplicatingTs
+
 		if event.Callback != nil {
 			callbacks = append(callbacks, event.Callback)
 		}
 
-		for _, row := range event.Event.Rows {
-			if len(startTs) == 0 || startTs[len(startTs)-1] != row.StartTs {
-				startTs = append(startTs, row.StartTs)
+		// Determine whether to use batch dml feature here.
+		if s.cfg.BatchDMLEnable && len(event.Event.Rows) > batchDMLThreshold {
+			tableColumns := firstRow.Columns
+			if firstRow.IsDelete() {
+				tableColumns = firstRow.PreColumns
+			}
+			// only use batch dml when the table has a handle key
+			if hasHandleKey(tableColumns) {
+				// TODO(dongmen): find a better way to get table info.
+				tableInfo := model.BuildTiDBTableInfo(tableColumns, firstRow.IndexColumns)
+				sql, value := batchSingleTxnDmls(event, tableInfo, translateToInsert)
+				sqls = append(sqls, sql...)
+				values = append(values, value...)
+				rowCount += len(sql)
```

Comment:
Maybe this variable is not accurate anymore. Should it be replaced with `statementCount`?